src/v/kafka/server/write_at_offset_stm.cc

```diff
@@ -0,0 +1,328 @@
+     */
+    auto u = co_await _sync_lock.get_units();
+    auto sync_result = co_await sync(timeout);
+    if (!sync_result) {
+        _inflight_last_offset.reset();
+        enqueued_promise.set_value();
+        co_return raft::errc::not_leader;
+    }
+
+    const auto stm_last_offset = expected_last_offset();
+    vlog(
+      _log.trace,
+      "Requested replicate at offset: {} with previous log offset: {}. "
+      "stm last offset: {} [inflight_last_offset: {}, last_offset: {}]",
+      expected_base_offset,
+      prev_log_offset,
+      stm_last_offset,
+      _inflight_last_offset,
+      _last_offset);
+
+    const auto effective_prev_log_offset = prev_log_offset.value_or(
+      kafka::prev_offset(expected_base_offset));
+    /*
+     * Write can only succeed if the effective_prev_log_offset ==
+     * stm_last_offset.
+     */
+    if (effective_prev_log_offset != stm_last_offset) {
+        enqueued_promise.set_value();
+        vlog(
+          _log.debug,
+          "Expected last log offset: {} does not match with last stm"
+          " tracked offset: {}",
+          effective_prev_log_offset,
+          stm_last_offset);
+        co_return make_error_code(errc::invalid_offset);
+    }
+
+    chunked_vector<model::record_batch> batches;
+    // fill the gap with ghost batches if required
+    if (effective_prev_log_offset != kafka::prev_offset(expected_base_offset)) {
+        // the term argument is not important here, it will be
+        // overriden in Raft layer
+        auto ghost_batches = model::make_ghost_batches(
+          kafka::offset_cast(kafka::next_offset(stm_last_offset)),
+          model::prev_offset(kafka::offset_cast(expected_base_offset)),
+          model::term_id{0});
+        batches.reserve(ghost_batches.size() + 1);
+        std::ranges::move(ghost_batches, std::back_inserter(batches));
+    }
+
+    _inflight_last_offset = term_offset{
+      .offset = kafka::offset(
+        expected_base_offset() + batch.header().last_offset_delta),
+      .in_sync_term = _insync_term};
+    batches.push_back(std::move(batch));
+    /**
+     * Write at offset state machine error handling is based on the assumption
+     * that if one of the replicate calls fails all the subsequent replicate
+     * calls in the same term will fail. In face of an error a new leader will
+     * call sync() waiting for the state from previous terms to be applied.
+     *
+     * Thanks to that assumption we can simply reset all inflight state instead
+     * of reverting to the previous last offset.
+     */
+    auto stages = try_replicate_in_stages(std::move(batches));
+
+    auto enq = std::move(stages.request_enqueued).finally([u = std::move(u)] {
+    });
+
+    std::move(enq).forward_to(std::move(enqueued_promise));
+
+    auto r_fut = co_await ss::coroutine::as_future(
+      std::move(stages.replicate_finished));
+    const bool needs_inflight_reset = inflight_last_offset_term_in_sync();
```

Comment:
I think this check is not tight enough... we have to cache _in_sync_term in after L106 and then compare with `inflight_last_offset->in_sync_term`, otherwise there could be a racy sync() that updates _insync_term in the background and this always true?